launch/config.py

```diff
@@ -0,0 +1,385 @@
+import copy
+import os
+
+import yaml
+
+import launch.util
+import test_util.aws
+from dcos_installer.config import config_sample
+from gen import get_dcosconfig_source_target_and_templates, stringify_configuration
+from gen.build_deploy.bash import onprem_source
+from gen.calc import validate_true_false
+from gen.internals import resolve_configuration, Scope, Source, Target, validate_one_of
+from pkgpanda.util import load_string, load_yaml
+from ssh.validate import source as ssh_source
+
+
+def expand_path(path: str, config_dir: str):
+    """ Returns absolute paths by performing ~ expansion or prepending the
+    config_dir to make an absolute path
+    """
+    if path.startswith('/'):
+        # nothing to do, return the path
+        return path
+    if path.startswith('~'):
+        # Try this and raise LauncherError
+        home_dir = os.environ['HOME']
+        return path.join(home_dir, path[1:])
+    # path is relative to config_dir
+    return os.path.join(config_dir, path)
+
+
+def get_validated_config(config_path: str):
+    """ Takes a dcos-launch configuration file and:
+    - string-ifies nested dicts
+    - expands paths into absolute paths, supporting tilde expansion
+    - converts provider_info_filename into provider_info
+    """
+    config_dir = os.path.dirname(os.path.abspath(config_path))
+    # Try and catch YamlParseError
+    config = load_yaml(config_path)
+    for k, v in config.items():
+        if k.endswith('_filename'):
+            config[k] = expand_path(v, config_dir)
+        if k == 'dcos_config':
+            for kk, vv in config[k].items():
+                if kk.endswith('_filename'):
+                    config[k][kk] = expand_path(vv, config_dir)
+        if isinstance(config[k], dict):
+            config[k] = yaml.dump(config[k])
+    config = stringify_configuration(config)
+    return validate_config(config)
+
+
+def get_user_argument_source(user_config: dict):
+    """ Unlike gen, nested config are allows here
+    """
+    user_source = Source(is_user=True)
+    for name, value in user_config.items():
+        user_source.add_must(name, value)
+    return user_source
+
+
+def validate_config(user_config: dict):
+    user_source = get_user_argument_source(user_config)
+    t = get_target()
+    sources = [get_source(), user_source]
+    resolver = resolve_configuration(sources, [t])
+    status = resolver.status_dict
+
+    if status['status'] == 'errors':
+        raise launch.util.LauncherError('ValidationError', pretty_print_validate_error(status))
+
+    late_variables = dict()
```

Comment:
@branden I copied this for completeness of validation, but I think I should just ban all late usage for this module. What do you think?